src/main/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtil.scala

```diff
@@ -44,12 +45,80 @@ trait CqlPreparedStatementUtil {
   def getParamsList(preparedStatement: PreparedStatement): List[Int]
 }
 
+object SessionCollectionResolver {
+  def get(session:Session, name:String):Option[Any] = session.attributes.get(name)
+
+  def getClz[T <: Any](session:Session, name:String):Option[Class[T]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case clz:Class[T] => Option(clz)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  def getIterable[T <: Any](session:Session, name:String):Option[lang.Iterable[T]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case rv:Iterable[T] => Option(rv.asJava)
+        case rv:lang.Iterable[T] => Option(rv)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  def getMap[K <: Any, V <: Any](session:Session, name:String):Option[util.Map[K,V]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case rv:Map[K,V] => Option(rv.asJava)
+        case rv:util.Map[K,V] => Option(rv)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  def getIterableClz[T <: Any](session:Session, name:String):Class[_ <: T] = {
```

Comment:
If you will apply my suggestions, then this method can be simplified to:
```
getClz(session, name + "-clz").getOrElse(() => {
      val iterator = getIterable(session, name)
      if (!iterator.hasNext) {
        throw new IllegalStateException("Iterable element class wasn't defined in Gatling session and Iterable is empty, cannot determine list type")
      }
      iterator.next.getClass
    })
```

Reply:
There are a few other cases to consider there re: the Option type, but your poin about getOrElse() is well taken. :)

Reply:
If you will follow my suggestion about don't use Optional<List> then there is no need for handling `Opitonal` - please try to simplify it and you will see that Option with List is not needed overhead.

Reply:
These patterns are very well-worn and established, but for sake of completeness let's follow through on the exercise.  I've changed getIterable to the following:

```scala
  def getIterable[T <: Any](session:Session, name:String):Iterable[T] = {
    get(session, name) match {
      case Some(rv: Iterable[T]@unchecked) => rv
      case Some(rv: lang.Iterable[T]@unchecked) => rv.asScala
      case _ => Iterable.empty
    }
  }
```

This forces me into something like the following for getIterableClz:

```scala
  def getIterableClz[T <: Any](session:Session, name:String):Class[_ <: T] = {
    val sessionOption:Option[Class[T]] = getClz(session, name + "-clz")
    sessionOption.getOrElse {

      val iterable:Iterable[T] = getIterable(session, name)
      // I now have no way to detect this case!
      //if (iterable.isEmpty) {
      //  throw new IllegalStateException("Iterable element class wasn't defined in Gatling session and Iterable is unavailable, cannot determine iterable type")
      //}
      if (iterable.isEmpty) {
        throw new IllegalStateException("Iterable element class wasn't defined in Gatling session and Iterable is empty, cannot determine iterable type")
      }
      iterable.head.getClass
    }
  }
```

If you can see a way to restore the check I've commented out I'm open to it, but with a lack of an option type that information is now essentially lost.

Perhaps you're arguing that we don't need these kind of granular exceptions.  If that's the case I'd still disagree; it seems useful to provide the user as much info as we can, and distinguishing between missing contents in a session and an empty collection in a session seems worthwhile.

Reply:
Do you think that this part of the message:`is unavailable` or `is empty` is valuable for clients of this code? (because I see that the rest of the message is exactly the same) I think that for the callers this distinction of error messages will not be seen and they will treat both of them the same way. Why would they need to handle those exceptions differently? Could you give me an example?
If they need to do it then we should introduce different kinds of exceptions that `IllegalStateException` because even if you will leave the current logic, the client to distinguish between the exceptions will need to do if logic on the content of the exception message - that seems far from ideal.

Reply:
It's unlikely that simulation code will be handling these exceptions.  The much more common case (I would expect) is that a simulation might forget to add something to the Gatling session that is subsequent referenced in a Gatling action.  In that case the message will be handed back to the user, so the message matters much more than the type here.

That's why I was so adamant about wanting to clearly distinguish between something missing from the session and something being present (but containing no values) in the session.  When a users Gatling run fails I want to give them as much detailed information as possible to help them diagnose the failure.